### PR TITLE
Do not upload files with identical etags

### PR DESCRIPTION
**Pull request by ddfreyne**
This is continued off #480.

This is quite tricky to write tests for. Not sure how to handle that.

### CODE REVIEW

**Comment by ddfreyne** [comment]
CC @nanoc/contributors — would like a +1

**Comment by mpapis** [comment]
yes looks good :+1:, I guess my only remark is that the run grows quite big, but it's not related to this PR

**Comment by ddfreyne** [comment]
Will look into adding some tests for this. (Will be hard, I guess!)

File: lib/nanoc/extra/deployers/fog.rb [hunk 1]
```diff
@@ -70,11 +71,17 @@ def run
         files = Dir['**/*'].select { |f| File.file?(f) }
         files.each do |file_path|
           key = path ? File.join(path, file_path) : file_path
+
+          keys_to_destroy.delete(key)
+
+          local_etag = calc_local_etag(file_path)
+          remote_etag = etags[key]
+          next if remote_etag && remote_etag == local_etag
+
```
Comment (mpapis) [inline lib/nanoc/extra/deployers/fog.rb]: avoid calculating local hash if remote is not available:
```ruby
remote_etag = etags[key]
if remote_etag
  local_etag = calc_local_etag(file_path)
  next if remote_etag == local_etag
end
```

**Comment by mpapis** [comment]
to simplify testing you could extract https://github.com/nanoc/nanoc/blob/aws-deploy-md5-v2/lib/nanoc/extra/deployers/fog.rb#L68-L86 to separate method `upload` - this also solves problem of growing `run` - and add `attr_accessor :directory, :etags, :keys_to_destroy` - this way you can check your custom `directory` and list of `etags` without actually connecting to aws.

File: lib/nanoc/extra/deployers/fog.rb (hunk 1 above)
Comment (ddfreyne) [inline lib/nanoc/extra/deployers/fog.rb]: Thanks, updated! (see #552—not this PR)

**Pull request closed without merging**